File: packages/nutzap-recovery/src/journal.ts
import { DatabaseSync } from 'node:sqlite';
import { chmodSync, closeSync, openSync } from 'node:fs';
import type { Event } from 'nostr-tools';
import { proofY, type Nutzap, type NutzapProof } from './protocol.js';
import type { PreparedRedemption, PreparedSpend, RedemptionRecord } from './types.js';

/** Private, disposable single-wallet journal. Transactions never span network calls. */
export class Journal {
  readonly #db: DatabaseSync;
  constructor(path: string) {
    closeSync(openSync(path, 'a', 0o600));
    chmodSync(path, 0o600);
    this.#db = new DatabaseSync(path);
    this.#db.exec('PRAGMA busy_timeout=5000; PRAGMA synchronous=FULL;');
    this.#db
      .exec(`CREATE TABLE IF NOT EXISTS redemptions(id TEXT PRIMARY KEY, record TEXT NOT NULL);
      CREATE TABLE IF NOT EXISTS inputs(mint TEXT, y TEXT, redemption TEXT NOT NULL, PRIMARY KEY(mint,y));
      CREATE TABLE IF NOT EXISTS outputs(mint TEXT, y TEXT, redemption TEXT NOT NULL, PRIMARY KEY(mint,y));
      CREATE TABLE IF NOT EXISTS acknowledgements(id TEXT, target TEXT, PRIMARY KEY(id,target));`);
  }
  get(id: string): RedemptionRecord | undefined {
    const row = this.#db.prepare('SELECT record FROM redemptions WHERE id=?').get(id);
    if (!row) return undefined;
    const record: RedemptionRecord = JSON.parse(String(row.record));
    record.published = this.#db
      .prepare('SELECT target FROM acknowledgements WHERE id=?')
      .all(id)
      .map((r) => String(r.target));
    return record;
  }
  reserve(zap: Nutzap, plan: PreparedRedemption): RedemptionRecord {
    this.#db.exec('BEGIN IMMEDIATE');
    try {
      let record = this.get(zap.id);
      if (record) {
        if (
          record.zap.recipient !== zap.recipient ||
          record.zap.lockingKey !== zap.lockingKey ||
          record.zap.amount !== zap.amount ||
          JSON.stringify(record.zap.proofs.map((p) => [proofY(p), p.amount, p.id, p.C]).sort()) !==
            JSON.stringify(zap.proofs.map((p) => [proofY(p), p.amount, p.id, p.C]).sort())
        )
          throw new Error('Conflicting nutzap economic identity');
      } else {
        for (const p of zap.proofs)
          this.#db.prepare('INSERT INTO inputs VALUES(?,?,?)').run(zap.mint, proofY(p), zap.id);
        record = { zap, plan, credit: null, events: [], published: [] };
        this.#db.prepare('INSERT INTO redemptions VALUES(?,?)').run(zap.id, JSON.stringify(record));
      }
      this.#db.exec('COMMIT');
      return record;
    } catch (error) {
      this.#db.exec('ROLLBACK');
      throw error;
    }
  }
  credit(
    id: string,
    amount: number,
    events: Event[],
    proofs: NutzapProof[],
    origin: 'local' | 'relay' = 'local',
  ): RedemptionRecord {
    this.#db.exec('BEGIN IMMEDIATE');
    try {
      const record = this.get(id);
      if (!record) throw Error('Missing prepared nutzap');
      if (record.credit === null) {
        // A second receipt must not turn the same bearer proofs into a second balance.
        for (const proof of proofs)
          this.#db
            .prepare('INSERT INTO outputs VALUES(?,?,?)')
            .run(record.zap.mint, proofY(proof), id);
        record.credit = amount;
        record.origin = origin;
        record.events = events;
        record.wallet = { token: events.find((e) => e.kind === 7375) ?? null, proofs, retired: [] };
        this.#db
          .prepare('UPDATE redemptions SET record=? WHERE id=?')
          .run(JSON.stringify(record), id);
      }
      this.#db.exec('COMMIT');
      return record;
    } catch (error) {
      this.#db.exec('ROLLBACK');
      throw error;
    }
  }
  acknowledge(id: string, target: string): void {
    this.#db.prepare('INSERT OR IGNORE INTO acknowledgements VALUES(?,?)').run(id, target);
  }
  #update(id: string, change: (record: RedemptionRecord) => void): RedemptionRecord {
    this.#db.exec('BEGIN IMMEDIATE');
    try {
      const record = this.get(id);
      if (!record || record.credit === null) throw Error('Missing credited nutzap');
      change(record);
      this.#db
        .prepare('UPDATE redemptions SET record=? WHERE id=?')
        .run(JSON.stringify(record), id);
      this.#db.exec('COMMIT');
      return record;
    } catch (error) {
      this.#db.exec('ROLLBACK');
      throw error;
    }
  }
  prepareSpend(id: string, amount: number, plan: PreparedSpend): RedemptionRecord {
    return this.#update(id, (record) => {
      if (record.spend) {
        if (record.spend.amount !== amount) throw Error('Conflicting spend amount');
        return;
      }
      if (!record.wallet?.token || record.wallet.retired.length)
        throw Error('Requires initial wallet token');
      record.spend = { amount, plan, events: [], sent: [] };
      // Prepared inputs are reserved, never advertised as spendable during recovery.
      record.wallet.proofs = [];
    });
  }
  #wallet(
    record: RedemptionRecord,
    token: Event | null,
    proofs: NutzapProof[],
    retired: string[],
  ): void {
    const tombstones = [...new Set([...(record.wallet?.retired ?? []), ...retired])];
    if (token && tombstones.includes(token.id)) throw Error('Retired wallet token');
    for (const proof of proofs) {
      const owner = this.#db
        .prepare('SELECT redemption FROM outputs WHERE mint=? AND y=?')
        .get(record.zap.mint, proofY(proof));
      if (owner && owner.redemption !== record.zap.id)
        throw Error('Proof already belongs to another receipt');
      this.#db
        .prepare('INSERT OR IGNORE INTO outputs VALUES(?,?,?)')
        .run(record.zap.mint, proofY(proof), record.zap.id);
    }
    record.wallet = { token, proofs, retired: tombstones };
  }
  wallet(id: string, token: Event | null, proofs: NutzapProof[], retired: string[]): boolean {
    let updated = false;
    this.#update(id, (record) => {
      if (record.spend && !record.spend.events.length) return;
      this.#wallet(record, token, proofs, retired);
      updated = true;
    });
    return updated;
  }
  finishSpend(
    id: string,
    events: Event[],
    keep: NutzapProof[],
    sent: NutzapProof[],
  ): RedemptionRecord {
    return this.#update(id, (record) => {
      if (!record.spend) throw Error('Missing prepared spend');
      if (record.spend.events.length) return;
      const original = record.events.find((e) => e.kind === 7375)!;
      this.#wallet(
        record,
        events.find((e) => e.kind === 7375)!,
        keep,
        [original.id],
      );
      record.spend.events = events;
      record.spend.sent = sent;
    });
  }
  summary(): { credits: number; balance: number } {
    const records = this.#db
      .prepare('SELECT record FROM redemptions')
      .all()
      .map((r) => JSON.parse(String(r.record)) as RedemptionRecord);
    return {
      credits: records.filter((r) => r.credit !== null && r.origin !== 'relay').length,
      balance: records.reduce(
        (sum, r) =>
          sum + (r.wallet ? r.wallet.proofs.reduce((n, p) => n + p.amount, 0) : (r.credit ?? 0)),
        0,
      ),
    };
  }
  close(): void {
    this.#db.close();
  }
}

File: adapters/cdk/src/nutzap/journal.rs
use super::{Record, Result};
use cdk::nuts::Proof;
use nostr::prelude::Event;
use rusqlite::{Connection, OptionalExtension, params};
use std::{
    fs::OpenOptions,
    os::unix::fs::{OpenOptionsExt, PermissionsExt},
    path::Path,
    time::Duration,
};

/// Private disposable journal. All economic changes are single SQLite transactions.
pub struct Journal(Connection);
impl Journal {
    pub fn open(path: &Path) -> Result<Self> {
        OpenOptions::new()
            .create(true)
            .append(true)
            .mode(0o600)
            .open(path)
            .map_err(|_| "journal_open_failed")?;
        std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o600))
            .map_err(|_| "journal_permissions_failed")?;
        let db = Connection::open(path).map_err(|_| "journal_open_failed")?;
        db.busy_timeout(Duration::from_secs(5))
            .map_err(|_| "journal_setup_failed")?;
        db.execute_batch("PRAGMA synchronous=FULL;
            CREATE TABLE IF NOT EXISTS redemptions(id TEXT PRIMARY KEY, record TEXT NOT NULL);
            CREATE TABLE IF NOT EXISTS inputs(mint TEXT NOT NULL,y TEXT NOT NULL,redemption TEXT NOT NULL,PRIMARY KEY(mint,y));
            CREATE TABLE IF NOT EXISTS outputs(mint TEXT NOT NULL,y TEXT NOT NULL,redemption TEXT NOT NULL,PRIMARY KEY(mint,y));
            CREATE TABLE IF NOT EXISTS acknowledgements(id TEXT NOT NULL,target TEXT NOT NULL,PRIMARY KEY(id,target));").map_err(|_| "journal_setup_failed")?;
        Ok(Self(db))
    }
    pub fn get(&self, id: &str) -> Result<Option<Record>> {
        let value: Option<String> = self
            .0
            .query_row("SELECT record FROM redemptions WHERE id=?", [id], |r| {
                r.get(0)
            })
            .optional()
            .map_err(|_| "journal_read_failed")?;
        value
            .map(|v| serde_json::from_str(&v).map_err(|_| "invalid_journal"))
            .transpose()
    }
    pub fn reserve(&mut self, record: &Record) -> Result<Record> {
        let tx = self
            .0
            .transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)
            .map_err(|_| "journal_busy")?;
        let existing: Option<String> = tx
            .query_row(
                "SELECT record FROM redemptions WHERE id=?",
                [&record.zap.id],
                |r| r.get(0),
            )
            .optional()
            .map_err(|_| "journal_read_failed")?;
        let saved = if let Some(value) = existing {
            let saved: Record = serde_json::from_str(&value).map_err(|_| "invalid_journal")?;
            if saved.zap.event.id != record.zap.event.id
                || saved.zap.mint != record.zap.mint
                || saved.zap.recipient != record.zap.recipient
            {
                return Err("conflicting_nutzap");
            }
            saved
        } else {
            for proof in &record.zap.proofs {
                tx.execute(
                    "INSERT INTO inputs VALUES(?,?,?)",
                    params![
                        record.zap.mint,
                        proof.y().map_err(|_| "invalid_proof")?.to_string(),
                        record.zap.id
                    ],
                )
                .map_err(|_| "duplicate_input")?;
            }
            tx.execute(
                "INSERT INTO redemptions VALUES(?,?)",
                params![
                    record.zap.id,
                    serde_json::to_string(record).map_err(|_| "invalid_journal")?
                ],
            )
            .map_err(|_| "journal_write_failed")?;
            record.clone()
        };
        tx.commit().map_err(|_| "journal_commit_failed")?;
        Ok(saved)
    }
    pub fn credit(
        &mut self,
        record: &Record,
        proofs: &[Proof],
        events: Vec<Event>,
        origin: &str,
    ) -> Result<Record> {
        let tx = self
            .0
            .transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)
            .map_err(|_| "journal_busy")?;
        let value: String = tx
            .query_row(
                "SELECT record FROM redemptions WHERE id=?",
                [&record.zap.id],
                |r| r.get(0),
            )
            .map_err(|_| "journal_read_failed")?;
        let mut saved: Record = serde_json::from_str(&value).map_err(|_| "invalid_journal")?;
        if saved.credit.is_none() {
            for proof in proofs {
                tx.execute(
                    "INSERT INTO outputs VALUES(?,?,?)",
                    params![
                        record.zap.mint,
                        proof.y().map_err(|_| "invalid_proof")?.to_string(),
                        record.zap.id
                    ],
                )
                .map_err(|_| "duplicate_output")?;
            }
            let value = super::protocol::amount(proofs)?;
            if value.checked_add(saved.plan.fee) != Some(saved.zap.amount) {
                return Err("value_mismatch");
            }
            saved.credit = Some(value);
            saved.wallet = Some(super::Wallet {
                token: events.iter().find(|e| e.kind.as_u16() == 7375).cloned(),
                proofs: proofs.to_vec(),
                retired: vec![],
            });
            saved.events = events;
            saved.origin = Some(origin.to_owned());
            tx.execute(
                "UPDATE redemptions SET record=? WHERE id=?",
                params![
                    serde_json::to_string(&saved).map_err(|_| "invalid_journal")?,
                    record.zap.id
                ],
            )
            .map_err(|_| "journal_write_failed")?;
        }
        tx.commit().map_err(|_| "journal_commit_failed")?;
        Ok(saved)
    }
    fn update(
        &mut self,
        id: &str,
        change: impl FnOnce(&rusqlite::Transaction<'_>, &mut Record) -> Result<()>,
    ) -> Result<Record> {
        let tx = self
            .0
            .transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)
            .map_err(|_| "journal_busy")?;
        let value: String = tx
            .query_row("SELECT record FROM redemptions WHERE id=?", [id], |r| {
                r.get(0)
            })
            .map_err(|_| "missing_receipt")?;
        let mut record: Record = serde_json::from_str(&value).map_err(|_| "invalid_journal")?;
        if record.credit.is_none() {
            return Err("missing_credit");
        }
        change(&tx, &mut record)?;
        tx.execute(
            "UPDATE redemptions SET record=? WHERE id=?",
            params![
                serde_json::to_string(&record).map_err(|_| "invalid_journal")?,
                id
            ],
        )
        .map_err(|_| "journal_write_failed")?;
        tx.commit().map_err(|_| "journal_commit_failed")?;
        Ok(record)
    }
    pub fn prepare_spend(
        &mut self,
        id: &str,
        amount: u64,
        plan: super::SpendPlan,
    ) -> Result<Record> {
        self.update(id, |_, record| {
            if let Some(spend) = &record.spend {
                return if spend.amount == amount {
                    Ok(())
                } else {
                    Err("conflicting_spend")
                };
            }
            let wallet = record.wallet.as_mut().ok_or("missing_wallet")?;
            if wallet.token.is_none() || !wallet.retired.is_empty() || wallet.proofs.is_empty() {
                return Err("requires_initial_wallet");
            }
            record.spend = Some(super::Spend {
                amount,
                plan,
                events: vec![],
                sent: vec![],
            });
            wallet.proofs.clear();
            Ok(())
        })
    }
    fn set_wallet(
        tx: &rusqlite::Transaction<'_>,
        record: &mut Record,
        token: Option<Event>,
        proofs: Vec<Proof>,
        retired: Vec<String>,
    ) -> Result<()> {
        let mut tombstones: std::collections::BTreeSet<String> = record
            .wallet
            .as_ref()
            .map(|w| w.retired.clone())
            .unwrap_or_default()
            .into_iter()
            .collect();
        tombstones.extend(retired);
        if token
            .as_ref()
            .is_some_and(|e| tombstones.contains(&e.id.to_hex()))
        {
            return Err("retired_token");
        }
        for proof in &proofs {
            let y = proof.y().map_err(|_| "invalid_proof")?.to_string();
            let owner: Option<String> = tx
                .query_row(
                    "SELECT redemption FROM outputs WHERE mint=? AND y=?",
                    params![record.zap.mint, y],
                    |r| r.get(0),
                )
                .optional()
                .map_err(|_| "journal_read_failed")?;
            if owner.is_some_and(|o| o != record.zap.id) {
                return Err("duplicate_output");
            }
            tx.execute(
                "INSERT OR IGNORE INTO outputs VALUES(?,?,?)",
                params![record.zap.mint, y, record.zap.id],
            )
            .map_err(|_| "journal_write_failed")?;
        }
        record.wallet = Some(super::Wallet {
            token,
            proofs,
            retired: tombstones.into_iter().collect(),
        });
        Ok(())
    }
    pub fn wallet(
        &mut self,
        id: &str,
        token: Option<Event>,
        proofs: Vec<Proof>,
        retired: Vec<String>,
    ) -> Result<bool> {
        let mut updated = false;
        self.update(id, |tx, record| {
            // Re-read under the write transaction: a spend may have begun during relay/mint I/O.
            if record.spend.as_ref().is_some_and(|s| s.events.is_empty()) {
                return Ok(());
            }
            Self::set_wallet(tx, record, token, proofs, retired)?;
            updated = true;
            Ok(())
        })?;
        Ok(updated)
    }
    pub fn finish_spend(
        &mut self,
        id: &str,
        events: Vec<Event>,
        keep: Vec<Proof>,
        sent: Vec<Proof>,
    ) -> Result<Record> {
        self.update(id, |tx, record| {
            let spend = record.spend.as_ref().ok_or("missing_spend")?;
            if !spend.events.is_empty() {
                return Ok(());
            }
            let original = record
                .events
                .iter()
                .find(|e| e.kind.as_u16() == 7375)
                .ok_or("missing_token")?
                .id
                .to_hex();
            let token = events
                .iter()
                .find(|e| e.kind.as_u16() == 7375)
                .cloned()
                .ok_or("missing_token")?;
            Self::set_wallet(tx, record, Some(token), keep, vec![original])?;
            let spend = record.spend.as_mut().ok_or("missing_spend")?;
            spend.events = events;
            spend.sent = sent;
            Ok(())
        })
    }
    pub fn acknowledged(&self, id: &str, target: &str) -> Result<bool> {
        self.0
            .query_row(
                "SELECT EXISTS(SELECT 1 FROM acknowledgements WHERE id=? AND target=?)",
                params![id, target],
                |r| r.get(0),
            )
            .map_err(|_| "journal_read_failed")
    }
    pub fn acknowledge(&self, id: &str, target: &str) -> Result<()> {
        self.0
            .execute(
                "INSERT OR IGNORE INTO acknowledgements VALUES(?,?)",
                params![id, target],
            )
            .map_err(|_| "journal_write_failed")?;
        Ok(())
    }
}
